R/write_output.R

```diff
@@ -0,0 +1,338 @@
+#' fit object and prepares a fact table containing unique date-time-parameter
+#' combinations for further merging.
+#'
+#' @param fit A Stan fit object with posterior estimates.
+#'
+#' @return A list containing two elements: `stan_draws` (the extracted draws in
+#' long format) and `fact_table` (a table of unique date-time-parameter
+#' combinations).
+#' @noRd
+extract_draws_from_fit <- function(fit) {
+  # Step 1: Extract unique date-time-parameter combinations
+  fact_table <- fit[["estimates"]][["samples"]][,
+    c("date", "time", "parameter"),
+    with = FALSE
+  ]
+  fact_table <- stats::na.omit(unique(fact_table))
+
+  # Step 1.1: Add corresponding 'obs_cases' rows for 'latent_cases' dates
+  obs_fact_table <- fact_table[
+    fact_table[["parameter"]] == "imputed_reports",
+  ]
+  # The EpiNow2 summary table has the variable `imputed_reports`
+  # for nowcast-corrected cases, but not `obs_reports` for right-
+  # truncated cases to compare to the observed data. We want both.
+  #
+  # The dates for `obs_reports` are the same as for `imputed_reports`
+  # (their differences are the nowcast correction + error structure).
+  # Get the dates for `obs_reports` by pulling out the `imputed_reports`
+  # dates and update the associated variable name in-place. Bind it back
+  # to the original fact table to have all desired variable-date combinations.
+  data.table::set(obs_fact_table, j = "parameter", value = factor(
+    obs_fact_table[["parameter"]],
+    levels = c("imputed_reports"),
+    labels = c("obs_reports")
+  ))
+
+  # Combine original fact_table with new 'obs_reports' rows
+  fact_table <- rbind(fact_table, obs_fact_table, fill = TRUE)
+  data.table::setnames(
+    fact_table,
+    old = c("parameter"),
+    new = c(".variable")
+  )
+
+  # Step 2: Extract desired parameters from the Stan object as posterior draws
+  stanfit <- fit[["estimates"]][["fit"]]
+  # Hacky workaround to avoid R CMD check NOTE on "no visible global binding"
+  # for variables in a dataframe evaluated via NSE. To use tidybayes, we need to
+  # use NSE, so giving these a global binding. The standard dplyr hacks
+  # (str2lang, .data prefix) didn't work here because it's not dplyr and we're
+  # not accessing a dataframe.
+  imputed_reports <- obs_reports <- R <- r <- time <- NULL # nolint
+  stan_draws <- tidybayes::gather_draws(
+    stanfit,
+    imputed_reports[time],
+    obs_reports[time],
+    R[time],
+    r[time],
+  ) |>
+    data.table::as.data.table()
+
+  return(list(stan_draws = stan_draws, fact_table = fact_table))
+}
+
+#' Post-process and merge posterior draws with a fact table.
+#'
+#' This function merges posterior draws with a fact table containing
+#' date-time-parameter combinations. It also standardizes parameter names and
+#' renames key columns.
+#'
+#' @param draws A data.table of posterior draws (either raw or summarized).
+#' @param fact_table A data.table of unique date-time-parameter combinations.
+#'
```

Comment:
Need to document `geo_value` `model` and `disease` here?

Reply:
https://github.com/CDCgov/cfa-epinow2-pipeline/issues/65